core/shared/src/main/scala/matryoshka/instances/fixedpoint/package.scala

```diff
@@ -205,28 +281,38 @@ package object fixedpoint {
       pf.lift ⋙ fromOption
   }
 
+  implicit def recursiveTEitherFoldable[T[_[_]]: RecursiveT]
+      : Foldable[λ[α => T[α \/ ?]]] =
+    new Foldable[λ[α => T[α \/ ?]]] {
+      def foldMap[A, B: Monoid](fa: T[A \/ ?])(f: A ⇒ B) =
+        fa.cata[B](_.leftMap(f).merge)
+
+      def foldRight[A, B](fa: T[A \/ ?], z: ⇒ B)(f: (A, ⇒ B) ⇒ B) =
+        fa.cata[B](_.leftMap(f(_, z)).merge)
+    }
+
   implicit val partialMonad: Monad[Partial] = new Monad[Partial] {
     def point[A](a: => A) = Partial.now(a)
 
     def bind[A, B](fa: Partial[A])(f: A => Partial[B]) =
-      fa.project.fold(f, l => Partial.later(bind(l)(f)))
+      fa.cata[Partial[B]](_.leftMap(f).merge)
   }
 
   implicit class PartialOps[A](self: Partial[A]) {
     def step: A \/ Partial[A] = self.project
 
     /** Returns `left` if the result was found within the given number of steps.
       */
-    def runFor(steps: Int): A \/ Partial[A] =
-      if (steps <= 0) step else step >>= (_.runFor(steps - 1))
+    def runFor[N](steps: N)(implicit N: Recursive.Aux[N, Option]): A \/ Partial[A] =
+      (steps, self).anaM[Partial[A]] {
+        case (r, p) => r.project.fold[(A \/ Partial[A]) \/ (A \/ (N, Partial[A]))](
+          p.project.left)(
+          prev => p.project.bimap(_.left, (prev, _).right))
+      }.map(_.step).merge
 
     /** Run to completion (if it completes).
       */
-    // TODO: Alternatively `self.cata[A](_.merge)`, but that's not tailrec.
-    @tailrec final def unsafePerformSync: A = self.project match {
-      case -\/(a) => a
-      case \/-(p) => p.unsafePerformSync
-    }
+    final def unsafePerformSync: A = self.cata[A](_.merge)
```

Comment:
Yeah – it’s being handled by the fixed-point. So, for `Cofree[A \/ ?, B]`, that’d be `Trampoline`, for `Nu`, it’d be the delayed call, etc. For `Mu` and `Fix`, well, you’re going to have trouble constructing a very partial `Mu[A \/ ?]` before you ever get around to running it 😆

But I’m not certain about this, so I should probably add a nice test or two (with a comment that it depends on your stack size).